app/src/main/java/org/greenstand/android/TreeTracker/messages/ChatViewModel.kt

```diff
@@ -1,142 +1,76 @@
 package org.greenstand.android.TreeTracker.messages
 
 
-import androidx.compose.runtime.mutableStateListOf
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import org.greenstand.android.TreeTracker.models.Users
 import org.greenstand.android.TreeTracker.models.messages.DirectMessage
+import org.greenstand.android.TreeTracker.models.messages.MessagesRepo
+import org.greenstand.android.TreeTracker.models.user.User
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.get
+import java.util.*
 
 
-class ConversationUiState(
-    val channelName: String,
-    val channelMembers: Int,
-    initialMessages: List<DirectMessage>
-) {
-    private val _messages: MutableList<DirectMessage> =
-        mutableStateListOf(*initialMessages.toTypedArray())
-    val messages: List<DirectMessage> = _messages
+data class ChatState(
+    val messages: Flow<List<DirectMessage>>? = null,
+    val draftText: String = "",
+    val currentUser: User? = null,
+)
+
+class ChatViewModel(
+    private val userId: Long,
+    private val otherChatIdentifier: String,
+    private val users: Users,
+    private val messagesRepo: MessagesRepo,
+) : ViewModel() {
+
+    private val _state = MutableLiveData<ChatState>()
+    val state: LiveData<ChatState> = _state
 
-    fun addMessage(msg: DirectMessage) {
-        _messages.add(0, msg) // Add to the beginning of the list
+    init {
+        viewModelScope.launch {
+            val currentUser = users.getUser(userId)
+            val messages = messagesRepo.getDirectMessages(currentUser!!.wallet, otherChatIdentifier)
```

Comment:
@Elforama the messages in state object is always null. Either this call returns null or something. I also try to check the database inspector but the message entity wasn't showing.

Reply:
You are not subscribing to the data from the flow.

It should be  
'
messages.repo.getDirectmessages(...)
.collect {
state = state.copy(messages = it, ...)

}
`

Reply:
Thanks it works now.